src/sbvr-api/control-flow.ts

```diff
@@ -0,0 +1,60 @@
+import * as _ from 'lodash'
+import * as Promise from 'bluebird'
+
+// TODO: We force the return type here to merge `Promise<U> | Promise<U[]>` into `Promise<U | U[]>,
+// which is equivalent for our uses but fixes issues where typing were being rejected
+export const liftP = <T, U>(fn: (v: T) => U | Promise<U>): (a: T | T[]) => Promise<U | U[]> => {
+	return (a: T | T[]) => {
+		if (_.isArray(a)) {
+			// This must not be a settle as if any operation fails in a cs
+			// we want to discard the whole
+			return Promise.mapSeries(a, fn)
+		} else {
+			return Promise.resolve(a).then(fn)
+		}
+	}
+}
+
+interface MappingFunction {
+	<T, U>(a: T[], fn: (v: T) => U | Promise<U>): Promise<Array<U | Error>>
+}
```

Comment:
You can just do 
```typescript
type MappingFunction = <T, U>(a: T[], fn: (v: T) => U | Promise<U>) => Promise<Array<U | Error>>

Reply:
You can, but I was under the impression that interfaces are preferable? @pimterry ?

Reply:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases suggests interfaces are mildly preferable, but personally I don't have any strong opinions on it for this kind of case. It does make a larger difference if you're defining an object type, since you can't extend type aliases, but for this I'm not sure it makes a lot of difference.

Reply:
> One difference is that interfaces create a new name that is used everywhere. Type aliases don’t create a new name — for instance, error messages won’t use the alias name.

That one seems like it could be pretty helpful at times (seeing massive long type defs like this has been a pita for me in the past, compared to if it used a name)

Reply:
cc @CameronDiver you may want to know this about interfaces vs type aliases if you're currently preferring type aliases

Reply:
Ah interesting. I would have considered them equivalent with type aliases being the succinct way of writing it. Thanks.